CNF-10412: Remove progress file from the prep-handler

sakhoury opened the pull request:
The progress file used to track and facilitate the various sub-stages of the prep-stage is removed. A new prep-stage worker goroutine is used instead.

Support is added to abort the prep-stage during any of the intermediate sub-steps (i.e. pulling the seed image, setting up the stateroot, precaching images).

/cc @jc-rh @browsell @sudomakeinstall2

sudomakeinstall2 commented on controllers/prep_handlers.go:
```diff
@@ -36,26 +42,14 @@ import (
-	if err := updateProgressFile(progressfile, "completed-seed-image-pull"); err != nil {
-		r.Log.Error(err, "failed to update progress file")
-		return result, err
+	ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade, imageListFile string,
```
this fucntion can be removed.

sudomakeinstall2 commented on controllers/prep_handlers.go:
```diff
@@ -327,28 +300,163 @@ func (r *ImageBasedUpgradeReconciler) SetupStateroot(ctx context.Context, ibu *l
-		return result, err
-	}
-	if err := r.SetupStateroot(ctx, ibu); err != nil {
+	ibu *lcav1alpha1.ImageBasedUpgrade) (success bool, err error) {
```
this function can be removed

sudomakeinstall2 commented on controllers/prep_handlers.go:
```diff
@@ -327,28 +300,163 @@ func (r *ImageBasedUpgradeReconciler) SetupStateroot(ctx context.Context, ibu *l
+	return
+}
+
+func (r *ImageBasedUpgradeReconciler) setPrepFailedStatus(ibu *lcav1alpha1.ImageBasedUpgrade) {
```
this fucntion does not need to be part of reconcilier

sudomakeinstall2 commented on controllers/prep_handlers.go:
```diff
@@ -327,28 +300,163 @@ func (r *ImageBasedUpgradeReconciler) SetupStateroot(ctx context.Context, ibu *l
+		ibu.Generation)
+}
+
+func (r *ImageBasedUpgradeReconciler) setPrepSucceededStatus(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) {
```
this function does not need to be part of reconciler. We can pass in the message.

sudomakeinstall2 commented:
Awesome
/lgtm

sudomakeinstall2 commented:
/lgtm

jc-rh commented on controllers/prep_handlers.go:
```diff
@@ -125,89 +108,67 @@ func readPrecachingList(imageListFile, clusterRegistry, seedRegistry string) (im
+	status, err = r.Precache.QueryJobStatus(ctx)
 	if err != nil {
-		r.Log.Error(err, "Failed to get precaching job status")
+		r.Log.Info("Failed to get precaching job status")
```
Do we really want to skip printing the actual error?

sakhoury replied:
There will be redundant logging of the error since the error is logged in line 465 (`r.Log.Error(err, "Prep stage failed with error")`) when the prepStageWorker function returns the error. Hence I removed the log.Error here

sakhoury replied:
I think we may want to create a separate PR in which we review redundant & duplicated logs in other places in an attempt to clean-up the LCA logs.

Missxiaoguo commented on controllers/prep_handlers.go:
```diff
@@ -342,29 +303,140 @@ func (r *ImageBasedUpgradeReconciler) SetupStateroot(ctx context.Context, ibu *l
-		if err := updateProgressFile(progressfile, "Failed"); err != nil {
-			r.Log.Error(err, "failed to update progress file")
-			return result, err
+func setPrepFailedStatus(ibu *lcav1alpha1.ImageBasedUpgrade) {
```
We have similiar funcs for upgrade status update in controllers/utils/conditions.go. Do we want to move these to that file to maintain them in one place?

jc-rh commented on controllers/ibu_controller.go:
```diff
@@ -65,6 +65,28 @@ type ImageBasedUpgradeReconciler struct {
 	ManifestClient  *clusterinfo.InfoClient
 	OstreeClient    ostreeclient.IClient
 	Ops             ops.Ops
+	PrepConfig      *PrepConfig
```
Not to block this PR:
- The type, maybe even the variable itself, is not specific to the Prep stage.
- Config feels like a wrong name for this? Task/job/step all sounds a little better?

sakhoury replied:
I didn't like the name `config` either, I will change it to `task` as job is a k8s jargon.

jc-rh commented:
/lgtm

Missxiaoguo commented on main/main.go:
```diff
@@ -165,6 +165,7 @@ func main() {
 		ManifestClient:  clusterinfo.NewClusterInfoClient(mgr.GetClient()),
 		OstreeClient:    ostreeClient,
 		Ops:             op,
+		PrepConfig:      &controllers.PrepConfig{Active: false, Cancel: nil},
```
nit: Seems like `Success` is not explicitly initialized here, but the default value is false if it's not initialized.

sakhoury replied:
`Success` is explicitly initialized to false in `handlePrep` when a new reconciliation for prep-stage commences. I was avoiding to bloat the parameter initialization here, but I can add it here.

Missxiaoguo commented:
/lgtm

jc-rh commented:
/approve

sudomakeinstall2 commented:
/lgtm

I just remembered that we need to set msg when prep fails. I can create a separate PR for that.

sakhoury commented:
> I just remembered that we need to set status msg when prep fails. We can create a separate PR for that.
> 
> /lgtm

@sudomakeinstall2 Do we need something more than this https://github.com/sakhoury/lifecycle-agent/blob/refactor-prep-handler/controllers/utils/conditions.go#L287-L300? I do set the prep failed on any error encountered during the prepStageWorker routine.

sakhoury commented:
/hold 

Updating prep failed condition

sudomakeinstall2 commented:
Yes I think we need something more than "Prep failed". So that we don't have to look at the log every time to see what went wrong. for example my prep is failing in status I only see "Prep failed" but in the log I see
`2023-12-12T17:13:02Z    ERROR   controllers.ImageBasedUpgrade   Failed to setupStateroot        {"error": "failed ostree admin os-init: error: Creating ostree/deploy/rhcos_4.14.3: File exists\n: exit status 1"}`
i think we need to add this to message.

I can add that if you are busy with another task.

sakhoury commented:
Thanks for your comment @sudomakeinstall2! I have updated the prep failed:
```yaml
...
status:
  conditions:
  - lastTransitionTime: “2023-12-12T19:31:15Z”
    message: In progress
    observedGeneration: 3
    reason: InProgress
    status: “False”
    type: Idle
  - lastTransitionTime: “2023-12-12T19:32:15Z”
    message: ‘Prep failed due to failed to get version from ClusterInfo: failed to
      read and decode ClusterInfo file: open /host/var/lib/containers/storage/overlay/6bcd2a9377883d001621c6d8460c6df6c41fb6ffcee6a1dcd7276c2414500ee5/merged/manifest.json:
      no such file or directory’
    observedGeneration: 3
    reason: Failed
    status: “False”
    type: PrepInProgress
  - lastTransitionTime: “2023-12-12T19:32:15Z”
    message: Prep failed
    observedGeneration: 3
    reason: Failed
    status: “False”
    type: PrepCompleted
  observedGeneration: 3
```
I have also updated the in-progress message to reflect the current sub-task progress. Here is a sample output:
```yaml
…
status:
 conditions:
 - lastTransitionTime: “2023-12-12T20:05:02Z”
  message: In progress
  observedGeneration: 5
  reason: InProgress
  status: “False”
  type: Idle
 - lastTransitionTime: “2023-12-12T20:05:02Z”
  message: Setting up stateroot
  observedGeneration: 5
  reason: InProgress
  status: “True”
  type: PrepInProgress
 observedGeneration: 5
…
status:
 conditions:
 - lastTransitionTime: “2023-12-12T20:05:02Z”
  message: In progress
  observedGeneration: 5
  reason: InProgress
  status: “False”
  type: Idle
 - lastTransitionTime: “2023-12-12T20:05:02Z”
  message: Waiting for precaching job to complete
  observedGeneration: 5
  reason: InProgress
  status: “True”
  type: PrepInProgress
 observedGeneration: 5
...
status:
 conditions:
 - lastTransitionTime: “2023-12-12T20:05:02Z”
  message: In progress
  observedGeneration: 5
  reason: InProgress
  status: “False”
  type: Idle
 - lastTransitionTime: “2023-12-12T20:05:02Z”
  message: ‘Precaching progress: total: 126 (pulled: 13, skipped: 12, failed: 0)’
  observedGeneration: 5
  reason: InProgress
  status: “True”
  type: PrepInProgress
 observedGeneration: 5
```

sakhoury commented:
/unhold

jc-rh commented:
/lgtm

jc-rh commented:
/approve
